### PR TITLE
Add automatic Jib support for Maven

### PR DESCRIPTION
**Pull request by chanseokoh**
Warning: I'm pretty new to Go.

If `pom.xml` exists and has Jib configured in it, the main automatically runs Jib. However, `Dockerfile` exists, it still takes the first precedence.

For Jib, it builds directly to GCR. It also overrides any configured credential helper to always use `docker-credential-gcloud`. Tested this only locally.

### CODE REVIEW

**Comment by jamesward** [comment]
Thanks for this contribution and getting a conversation going around how to handle this.  Instead of running `mvn` as part of the button, I think it'd be preferable to have a `build` script like we were discussing in #85.  Because then we can easily support any project-based process that can create a container image.  And then we wouldn't need special handling for things like `mvnw` as well.  I have a PR for something pretty similar:
https://github.com/GoogleCloudPlatform/cloud-run-button/pull/86

But I'm awaiting review, so it could be totally wrong cause I don't really know what I'm doing.  :)

**Comment by ahmetb** [comment]
I'm in favor of adding a few more "auto-build" modes –but it should not conflict with `pack` support. This one seems mostly harmless (I'll review later once I'm back from OOO).

If we can satisfy Java repos without having to write a build script, that would be most favorable.

**Comment by ahmetb** [comment]
> Tested this only locally.

you can test it by building/pushing a **public** docker image + editing the Open in Cloud Shell URL we provide in readme with that image.

**Comment by jamesward** [comment]
Right now the detection is based on running the Maven build and checking for the jib plugin.  Which I think it the root of the problem for pack fallback.  So for instance, we'd do this if there is a `pom.xml`:
1. Run the maven build to check for jib (and ideally also support mvnw which downloads maven on-the-fly instead of using the Cloud Shell provided one)
2. If jib plugin doesn't exist, run pack

The overhead of this is potentially significant (like a minute) which I why I favor an explicit build script instead of the detection attempt.

**Comment by ahmetb** [comment]
Since I don’t know much Java: do we have to do maven build to “check for jib”? Isnt there more a static analysis we can do?

**Comment by jamesward** [comment]
We can grep for it. :)  But there are probably a million corner cases.  I don't think it's too much hassle to have someone add their build command to an `app.json` and then we can keep that concern out of the button.  I don't think we want to head down the path of having all of the detections for all the ways a project can produce a container, in our code.

**Comment by ahmetb** [comment]
Adding command to a script file is usually better than adding it to app.json. 
You almost never wanna deal with scripting semantics in a json/yaml file.

If we somehow said buildMode: jib explicitly in app.json, that would do fine.

**Comment by jamesward** [comment]
We'd still have to figure out how to run the build, is it Maven or Gradle?  Does it use mvnw or gradlew?    And then we have to add another build mode for sbt's docker builder (like jib but different).  Repeat for all the ways that someone can create a container from their build.  It's a lot of work and a lot to maintain to avoid allowing the user to specify their build command.

Yeah, commands in JSON aren't great but they are pretty common practice and there is already precedent with app.json's postdeploy command support.  If we don't want to allow commands in app.json we could alternatively only support file references in the app.json scripts.

**Comment by chanseokoh** [comment]
So I was also concerned about the slow Jib detection, which is due to Maven having to download a lot of artifacts into `~/.m2`. I think we can improve this by first doing a simple grep on pom.xml to see if `jib-maven-plugin` is configured, and if so, do the Maven detection. If we see `jib-maven-plugin`, there is a very good chance Jib is correctly configured in the default settings. The time spent on downloading the artifacts during detection won't be lost, because those artifacts would have to be downloaded anyway in the main Jib build later.

Personally, I hope the Cloud Run button does some special auto-handling for Jib projects so that repo owners can add the button easily without adding `app.json`. I also have thought about adding another URL parameter like `--disableJibDetection`. But having an extension to run arbitrary commands could be handy regardless.

**Comment by chanseokoh** [comment]
Added a quick peek check to grep `<artifactId>jib-maven-plugin</artifactId>` in `pom.xml`. Projects without Jib won't get slowed down. Also nothing to lose for Jib projects, as I explained.

File: cmd/cloudshell_open/jib.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		if (stat.Mode() & 0111) != 0 {
+			if wrapper, err := filepath.Abs(filepath.Join(dir, "mvnw")); err == nil {
+				executable = wrapper
+			}
+		}
+	}
+
+	cmd := exec.Command(executable, args...)
+	cmd.Dir = dir
+	return *cmd
```
Comment (ahmetb) [inline cmd/cloudshell_open/jib.go]: Try not to dereference a normally pointer type. Just return *Cmd from here

File: cmd/cloudshell_open/main.go [hunk 2]
```diff
@@ -231,33 +232,49 @@ func run(c *cli.Context) error {
 		fmt.Println(infoPrefix + " FYI, running the following command:")
 		cmdColor.Printf("\tdocker build -t %s %s\n", parameter(image), parameter("."))
 	} else {
-		fmt.Println(infoPrefix + " Attempting to build this application with Cloud Native Buildpacks (buildpacks.io)...")
-		fmt.Println(infoPrefix + " FYI, running the following command:")
-		cmdColor.Printf("\tpack build %s --path %s --builder heroku/buildpacks\n", parameter(image), parameter(appDir))
+		jibMaven, err = jibMavenConfigured(appDir)
+		if err != nil {
+			return err
```
Comment (ahmetb) [inline cmd/cloudshell_open/main.go]: wrap with a message explaining it failed to detect jib

File: cmd/cloudshell_open/main.go [hunk 3]
```diff
@@ -234,7 +234,7 @@ func run(c *cli.Context) error {
 	} else {
 		jibMaven, err = jibMavenConfigured(appDir)
 		if err != nil {
-			return err
+			return fmt.Errorf("filed to check if Jib is configured: %s", err)
```
Comment (ahmetb) [inline cmd/cloudshell_open/main.go]: failed*

File: cmd/cloudshell_open/jib.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+			}
+		}
+	}
+
+	cmd := exec.Command(executable, args...)
+	cmd.Dir = dir
+	return cmd
+}
+
+func fileHasString(filePath string, pattern string) (bool, error) {
```
Comment (ahmetb) [inline cmd/cloudshell_open/jib.go]: optional but this, seems unnecessary, you can just unfold into the call site.

File: cmd/cloudshell_open/jib.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	cmd := createMavenCommand(dir, "--batch-mode", "-Dmaven.test.skip=true",
+		"package", "jib:build", "-Dimage="+image, "-Djib.to.auth.credHelper=gcloud")
+	if b, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("Jib Maven build failed: %v, output:\n%s", err, string(b))
+	}
+	return nil
+}
+
+func jibMavenConfigured(dir string) (bool, error) {
+	if _, err := os.Stat(filepath.Join(dir, "pom.xml")); err != nil {
```
Comment (ahmetb) [inline cmd/cloudshell_open/jib.go]: if you actually unfold ioutil.ReadAll here, you can do os.IsNotExist() directly on its err, and therefore eliminate the os.Stat

**Comment by jamesward** [comment]
I think before we merge this approach we do need to see how this is going to all play out with mvnw, gradle, and gradlew support.  Cause we need to see how the UX comes together.  We don't want to have a bunch of messages for things that don't pertain to most users.  e.g. I don't think we should tell every user that falls through to buildpacks that we couldn't detect a Jib project since I think that gets confusing in many cases.  And yeah, we'll need to do some testing since we don't have integration tests that validate the right decision flow (e.g. falling back to buildpacks).

**Comment by chanseokoh** [comment]
> I don't think we should tell every user that falls through to buildpacks that we couldn't detect a Jib project 

I agree. If Jib is not detected, nothing about Jib will be logged–complete silence about Jib.
```
[ ? ] Choose a region to deploy this application: us-central1                                                       
[ ! ] Attempting to build this application with its Dockerfile...                                                                                                                                  
[ ! ] FYI, running the following command:                                                                           
        docker build -t gcr.io/chanseok-playground-new/cloud-run-hello .                                                          
[ ✓ ] Built container image gcr.io/chanseok-playground-new/cloud-run-hello                                                   
```

The repos I've been testing with are

1.  `--repo_url=https://github.com/GoogleCloudPlatform/cloud-run-hello.git`
2. `--repo_url=https://github.com/chanseokoh/jib-war-helloworld-test.git`
   - WAR project
3. `--repo_url=https://github.com/GoogleContainerTools/jib.git --dir=examples/dropwizard`
   - Official Jib repo. The example is under `examples/dropwizard`. The intended endpoint is at `...:8080/hello-world`.

> I think before we merge this approach we do need to see how this is going to all play out with mvnw, gradle, and gradlew support.

We can have more discussions, but what I've been thinking is that, we do the Jib Maven detection first and then Gradle. (Most repos will have either one anyway, unless they are primarily for examples.) Detecting and using a Maven wrapper is easy to support and already works in this PR. Gradle is no different in this regard. We've been supporting wrappers in Skaffold-Jib integration.

**Comment by jamesward** [comment]
Cool. Didn't see the mvnw support.  That is good.  I'll give this a whirl on some test repos and see how it goes.

File: cmd/cloudshell_open/jib.go (hunk 5 above)
Comment (chanseokoh) [reply cmd/cloudshell_open/jib.go]: Cool. I've simplified the code around this.

**Comment by chanseokoh** [comment]
> please also find a way to document this feature in README.md then we're good to go.

I'll try to think of something later after adding Jib Gradle support.

**Comment by chanseokoh** [comment]
Move forward with this?

**Comment by jamesward** [comment]
Sorry for the delay.  I'll get this reviewed asap.

**Comment by jamesward** [comment]
So sorry for the delay.  Finally getting a chance to try this and so far it's been great.

I can't remember if it was supposed to work, but the Gradle repo I tried didn't work with jib:
https://github.com/jamesward/hello-play-gradle

**Comment by chanseokoh** [comment]
This PR doesn't do anything about Gradle. Another PR later, when I get to that.

**Pull request merged**